### PR TITLE
WIP: Add new macros $DESTIP, $DESTPORT & $PROTO

### PR DESCRIPTION
**Pull request by bazsi**
This branch implements $DESTIP, $DESTPORT and $PROTO macros as discussed on Splunk's slack, https://splunk-usergroups.slack.com/archives/C4SL594CS/p1547051419025900

It has become a lot more complex than I would have liked, so the branch contains a number of
refactors.

It seems that the splunk channel is not on on a paid channel, so there's no history :( I didn't expect that. Anyway the use-case is as follows:
* someone has an appliance which sends out log messages via both UDP and TCP
* the format of the two are different, and he wants to capture either with the simplest possible filter
* netmask() doesn't work because the IP addresses are the same
* host() doesn't work because the hostnames are the same

That's how the idea of $PROTO was born. Then I've figured that if we have SOURCEIP and PROTO, we should have $DESTIP and $DESTPORT so we can have access the entire IP level tuple. Now as I see it, we don't yet have $SOURCEPORT, I can add that as well, the information is already available.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logreader.c [hunk 1]
```diff
@@ -399,6 +400,7 @@ log_reader_handle_line(LogReader *self, const guchar *line, gint length, LogTran
                   &self->options->parse_options);
 
   log_msg_set_saddr(m, aux->peer_addr ? : self->peer_addr);
+  log_msg_set_daddr(m, aux->local_addr ? : self->local_addr);
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/logreader.c]: I was just following the code that was already there.

Looking further, there are 22 instances of the same extension in the codebase. Honestly I wasn't aware this was an extension, but the earliest instance goes back in 2013, but I didn't go back through file renames and stuff.

These are the ones I've found:
```
a67ad5b6d8 (Balazs Scheidler 2018-03-26 06:40:34 +0200  59)              self->filename ? : "#buffer",
e326f5c796 (Balazs Scheidler 2013-05-21 11:54:54 +0200 258)               if (!_append_value(self, value ? : "", error))
2eb6731ece lib/cfg-tree.c (Balazs Scheidler  2016-09-29 22:49:47 +0200  173)           g_snprintf(buf, buf_len, "%s:%d:%d", self->filename ? : "#buffer", node->line, node->column);
d336a8950e lib/cfg-tree.c (Noémi Ványi       2016-08-29 11:27:42 +0200 1171)                  node->name ? : "#unnamed",
eaa14391c8 lib/cfg.c (Balazs Scheidler  2017-09-18 15:37:11 +0200 191)   gboolean autoload_enabled = atoi(cfg_args_get(self->globals, "autoload-compiled-modules") ? : "1");
0775a3b906 (Balazs Scheidler 2015-02-07 07:30:35 +0100 271)   if (!g_shell_parse_argv(self->command_buffer ? : "", &argc, &argv, &error))
55d3395155 lib/logreader.c (Balazs Scheidler 2019-01-14 18:32:14 +0100 402)   log_msg_set_saddr(m, aux->peer_addr ? : self->peer_addr);
0ea54517c0 lib/logreader.c (Balazs Scheidler 2019-01-14 19:32:04 +0100 403)   log_msg_set_daddr(m, aux->local_addr ? : self->local_addr);
476f8147fd lib/plugin.c (Balazs Scheidler 2015-11-22 11:36:00 +0100 285)   module_path_dirs = g_strsplit(module_path ? : "", G_SEARCHPATH_SEPARATOR_S, 0);
1980bde933 lib/plugin.c (Balazs Scheidler 2012-03-04 14:04:39 +0100 385)       module_init_func = plugin_get_module_init_name(module_info->canonical_name ? : module_name);
f9f32c61c4 lib/plugin.c (Antal Nemes      2018-04-26 09:11:21 +0200 430)   mod_paths = g_strsplit(context->module_path ? : "", G_SEARCHPATH_SEPARATOR_S, 0);
43007c54e0 lib/scanner/kv-scanner/kv-scanner.c (Balazs Scheidler 2017-12-01 16:57:02 -0700 341)   self->pair_separator = pair_separator ? : ", ";
5434034642 lib/stats/stats-cluster.c (Laszlo Budai     2017-04-26 14:43:48 +0200  33)   dst->id = g_strdup(src->id ? : "");
5434034642 lib/stats/stats-cluster.c (Laszlo Budai     2017-04-26 14:43:48 +0200  34)   dst->instance = g_strdup(src->instance ? : "");
de1ef3c5c1 lib/tlscontext.c              (Balazs Scheidler 2018-04-12 15:59:47 +0200  766)   self->location = g_strdup(location ? : "n/a");
3ee7a04aef lib/value-pairs/tests/test_value_pairs.c (Balazs Scheidler 2018-09-15 06:08:51 +0200 149)       ValuePairsTransformSet *vpts = value_pairs_transform_set_new(transformed_keys ? : "*");
3ee7a04aef lib/value-pairs/tests/test_value_pairs.c (Balazs Scheidler 2018-09-15 06:08:51 +0200 162)                    scope, transformed_keys ? : "*", vp_keys->str, expected);
3ef7c6d1b0 (Balazs Scheidler 2017-12-01 10:25:48 -0800 140)   config = g_strdup_printf(config_fmt, topic, varargs ? : "");
573376ba05 (Balazs Scheidler 2015-11-22 18:56:44 +0100 256)   invalid_chars = invalid_chars ? : g_strdup("/");
573376ba05 (Balazs Scheidler 2015-11-22 18:56:44 +0100 257)   replacement = replacement ? : g_strdup("_");
0804b2d80f modules/dbparser/tests/test_radix.c (norberttakacs    2018-08-28 07:57:34 +0200   57)   r_insert_node(root, dup, value ? : (gpointer)key, NULL);
109e13195f (Balazs Scheidler 2018-03-19 17:12:32 +0100  68)   res = PyObject_CallFunction(print_exception, "OOO", exc, value, tb ? : Py_None);

```

I don't seem to be the only one using it :)

With that I think this request is moot, at the very least this should be a separate initiative, if we decide that "? :" is something we shouldn't use. And then if we do, we should remove all instances, not just this.

File: lib/logmsg/logmsg.h [hunk 2]
```diff
@@ -200,7 +200,16 @@ struct _LogMessage
...
+         recursed:1,
+
+         /* NOTE: proto is just 6 bits wide, but with that it fills a hole
+          * not taking any tolls on the structure size.  Realistically, we'd
+          * be storing IPPROTO_UDP and TCP in there, which fits the 6 bits.
+          * This is closely related to saddr/daddr and indicates the IP
+          * protocol that was used to deliver the datagram carrying this
+          * LogMessage.  */
+
+         proto:6;
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/logmsg/logmsg.h]: I know, but even then, I was trying to use the same name as the macro name for the user. It was originally ip_protocol in my code, but in the macro names, we have:

$SOURCEIP
$DESTIP
$DESTPORT

In that line, I feel that $PROTO is a more natural name. ip_protocol is used as an argument for our socket based drivers, but that specifies ipv4 or ipv6, not the transport protocol that runs on top. Wtih ip_protocol being ipv4/ipv6, that would become "protocol" that's what I have shortened to "proto".

Anyhow, this was my reasoning above, alternative suggestions appreciated.

File: lib/logmsg/logmsg.h [hunk 3]
```diff
@@ -170,10 +170,15 @@ struct _LogMessage
 
   gint ack_and_ref_and_abort_and_suspended;
 
+  /* NOTE: in theory this should be a size_t (or gsize), however that takes
+   * 8 bytes, and it's highly unlikely that we'd be using more than 4GB for
+   * a LogMessage */
+
+  gint allocated_bytes;
```
[the remark this answers is not in the document]
Comment (bazsi) [reply lib/logmsg/logmsg.h]: good catch, fixed.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logreader.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (lbudai) [reply lib/logreader.c]: indeed... (and yes, based on the logs, I also used this extension :) ).

File: lib/template/macros.c [hunk 4]
```diff
@@ -190,6 +190,9 @@ LogMacroDef macros[] =
   { "SDATA", M_SDATA },
   { "MSGHDR", M_MSGHDR },
   { "SOURCEIP", M_SOURCE_IP },
+  { "DESTIP", M_DEST_IP },
+  { "DESTPORT", M_DEST_PORT },
+  { "PROTO", M_PROTO },
```
Comment (furiel) [inline lib/template/macros.c]: Checked an example message with an `stdin` source, and these macros are filled,
```
"DESTPORT":"0"
"DESTIP":"127.0.0.1"
"PROTO":"0"
```
though these are network source specific macros.

Maybe we can reason for `DESTIP`, the way `SOURCEIP` exists.
`DESTPORT` and `PROTO` still looks strange. On the upside, zero for port is not valid value.
On the other hand, the zero value for proto is a valid protocol number for `HOPOPT`.

I was thinking if we would be better off with empty values, but that can also create confusion: users need to use `--omit-empty-values` (once that PR is merged) to exclude these values. Which is not easy remember to look out for.

**Comment by bazsi** [comment]
On Mon, Feb 4, 2019 at 8:37 AM furiel <notifications@github.com> wrote:

> *@furiel* commented on this pull request.
> ------------------------------
>
> In lib/template/macros.c
> <https://github.com/balabit/syslog-ng/pull/2495#discussion_r253366808>:
>
> > @@ -190,6 +190,9 @@ LogMacroDef macros[] =
>    { "SDATA", M_SDATA },
>    { "MSGHDR", M_MSGHDR },
>    { "SOURCEIP", M_SOURCE_IP },
> +  { "DESTIP", M_DEST_IP },
> +  { "DESTPORT", M_DEST_PORT },
> +  { "PROTO", M_PROTO },
>
> Checked an example message with an stdin source, and these macros are
> filled,
>
> "DESTPORT":"0"
> "DESTIP":"127.0.0.1"
> "PROTO":"0"
>
> though these are network source specific macros.
>
> Maybe we can reason for DESTIP, the way SOURCEIP exists.
>

btw, In this case SOURCEIP would be 127.0.0.1 as well.


> DESTPORT and PROTO still looks strange. On the upside, zero for port is
> not valid value.
> On the other hand, the zero value for proto is a valid protocol number for
> HOPOPT.
>
hopopt is not a valid protocol number, it's only used as an identifier for
an IPv6 extension header, as described here:
http://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml


> I was thinking if we would be better off with empty values, but that can
> also create confusion: users need to use --omit-empty-values (once that
> PR is merged) to exclude these values. Which is not easy remember to look
> out for.
>
I don't think we are better off with empty values. Also, these fields
wouldn't be included in outputs, unless someone explicitly wants them.
Using something like `$(format-json *)` does not make sense even today, as
there's lot of redundancy in its output. This leads us to that
`$(format-json *)` shouldn't behave as it does today, but that's a
different discussion.

Like I mentioned, $PROTO was the initial trigger of the feature, which I
couldn't resolve without adding it as $SOURCE is not good enough if you
have multiple drivers in the same source statement.

Using ip:port+protocol as a unique identifier makes sense on the other
hand. It should uniquely identify the port number in case we have multiple
listeners, even if they use a different log transport mechanism, as each
will need to open a socket eventually.

**Comment by furiel** [comment]
I am not particularly against introducing the triplet. My concern is that it is coupled to the message structure, even when there no context for it, for example in case of file source, generator source, or internal source. They all will have these information default to something, although it is not relevant for them.

Similarly, with the `LogReader` class, that is used also for file source for example, starts to handle ip protocol specific information, and that's what looks odd to me.

Now I am wild-guessing so I might be much off track, but me seems storing the triplet in the nvtable would probably equally convenient for the users as providing these information as macros. Maybe these could be passed around into `LogReader` using `LogTransportAuxData`.

**Comment by furiel** [comment]
@bazsi could you resolve the conflicts?

**Comment by bazsi** [comment]
This just came up as a second request, so I am trying to rebase and then reopen this.

**Pull request closed without merging**